### ISSUE TITLE
TransferFunction order reduction

### ISSUE THREAD

**Issue by moorepants** [sympy/sympy#26160]
Here is a simple feedback system I was working with:

```python
import sympy as sm
import sympy.physics.control as cn

Ib, Is, c, m, h, g, l2, l1 = sm.symbols('I_b, I_s, c, m, h, g, l2, l1',
                                        real=True, nonnegative=True)
KD, KP, v = sm.symbols('K_D, K_P, v', real=True)
s = sm.symbols('s')

tau1_sq = (Ib + m*h**2)/m/g/h
tau2 = l2/v
tau3 = v/(l1 + l2)
K = v**2/g/(l1 + l2)

Gtheta = cn.TransferFunction(-K*(tau2*s + 1), tau1_sq*s**2 - 1, s)
Gdelta = cn.TransferFunction(1, Is*s**2 + c*s, s)
Dcont = cn.TransferFunction(KD*s, 1, s)

Ginner = cn.Feedback(Dcont*Gdelta, Gtheta)
```

The numerator and denominator polynomials take this form:

```
In [38]: sm.pprint(Ginner.doit().expand().num.collect(s))
 5 ⎛                                                  2                    2       ⎞    4 ⎛                                               2                   2       ⎞    3 ⎛          2   
s ⋅⎝I_b⋅Iₛ⋅K_D⋅g⋅l₁⋅v + I_b⋅Iₛ⋅K_D⋅g⋅l₂⋅v + Iₛ⋅K_D⋅g⋅h ⋅l₁⋅m⋅v + Iₛ⋅K_D⋅g⋅h ⋅l₂⋅m⋅v⎠ + s ⋅⎝I_b⋅K_D⋅c⋅g⋅l₁⋅v + I_b⋅K_D⋅c⋅g⋅l₂⋅v + K_D⋅c⋅g⋅h ⋅l₁⋅m⋅v + K_D⋅c⋅g⋅h ⋅l₂⋅m⋅v⎠ + s ⋅⎝- Iₛ⋅K_D⋅g ⋅h⋅

                 2         ⎞    2 ⎛         2                   2         ⎞
l₁⋅m⋅v - Iₛ⋅K_D⋅g ⋅h⋅l₂⋅m⋅v⎠ + s ⋅⎝- K_D⋅c⋅g ⋅h⋅l₁⋅m⋅v - K_D⋅c⋅g ⋅h⋅l₂⋅m⋅v⎠

In [39]: sm.pprint(Ginner.doit().expand().den.collect(s))
 6 ⎛      2                2            2    2            2    2       ⎞    5 ⎛                                                  2                    2       ⎞    4 ⎛     2               2
s ⋅⎝I_b⋅Iₛ ⋅g⋅l₁⋅v + I_b⋅Iₛ ⋅g⋅l₂⋅v + Iₛ ⋅g⋅h ⋅l₁⋅m⋅v + Iₛ ⋅g⋅h ⋅l₂⋅m⋅v⎠ + s ⋅⎝2⋅I_b⋅Iₛ⋅c⋅g⋅l₁⋅v + 2⋅I_b⋅Iₛ⋅c⋅g⋅l₂⋅v + 2⋅Iₛ⋅c⋅g⋅h ⋅l₁⋅m⋅v + 2⋅Iₛ⋅c⋅g⋅h ⋅l₂⋅m⋅v⎠ + s ⋅⎝I_b⋅c ⋅g⋅l₁⋅v + I_b⋅c 

            2  2              2  2                             2    2    2           2    2       ⎞    3 ⎛                3           2                    2                            2⎞  
⋅g⋅l₂⋅v - Iₛ ⋅g ⋅h⋅l₁⋅m⋅v - Iₛ ⋅g ⋅h⋅l₂⋅m⋅v - Iₛ⋅K_D⋅g⋅h⋅l₂⋅m⋅v  + c ⋅g⋅h ⋅l₁⋅m⋅v + c ⋅g⋅h ⋅l₂⋅m⋅v⎠ + s ⋅⎝- Iₛ⋅K_D⋅g⋅h⋅m⋅v  - 2⋅Iₛ⋅c⋅g ⋅h⋅l₁⋅m⋅v - 2⋅Iₛ⋅c⋅g ⋅h⋅l₂⋅m⋅v - K_D⋅c⋅g⋅h⋅l₂⋅m⋅v ⎠ +

  2 ⎛               3    2  2             2  2         ⎞
 s ⋅⎝- K_D⋅c⋅g⋅h⋅m⋅v  - c ⋅g ⋅h⋅l₁⋅m⋅v - c ⋅g ⋅h⋅l₂⋅m⋅v⎠
```

It would likely be preferable for both the numerator and denominator to be automatically divided by `s**2` to reduce the order of the polynomials to 3 (numerator) and 4 (denominator). Or there should be a method that does this.

The generic application of simplify gives the desired effect, but this may not always do so (not sure) and there is a very specific operation to reduce the order that is likely more efficient.

Here is the result with simplify:

```
In [45]: sm.pprint(Ginner.doit().simplify().expand().num.collect(s))
                               2 ⎛                                2             2     ⎞
K_D⋅g⋅h⋅l₁⋅m + K_D⋅g⋅h⋅l₂⋅m + s ⋅⎝-I_b⋅K_D⋅l₁ - I_b⋅K_D⋅l₂ - K_D⋅h ⋅l₁⋅m - K_D⋅h ⋅l₂⋅m⎠

In [46]: sm.pprint(Ginner.doit().simplify().expand().den.collect(s))
         2                              3 ⎛                             2            2     ⎞    2 ⎛                          2           2     ⎞                                            
K_D⋅h⋅m⋅v  + c⋅g⋅h⋅l₁⋅m + c⋅g⋅h⋅l₂⋅m + s ⋅⎝-I_b⋅Iₛ⋅l₁ - I_b⋅Iₛ⋅l₂ - Iₛ⋅h ⋅l₁⋅m - Iₛ⋅h ⋅l₂⋅m⎠ + s ⋅⎝-I_b⋅c⋅l₁ - I_b⋅c⋅l₂ - c⋅h ⋅l₁⋅m - c⋅h ⋅l₂⋅m⎠ + s⋅(Iₛ⋅g⋅h⋅l₁⋅m + Iₛ⋅g⋅h⋅l₂⋅m + K_D⋅h⋅l₂⋅m

   
⋅v)
```

**Comment by moorepants** [comment]
Related: https://github.com/sympy/sympy/issues/25340

**Comment by oscarbenjamin** [comment]
This is a weird form:
```
In [14]: Ginner
Out[14]: 
             ⎛K_D⋅s⎞ ⎛     1     ⎞              
             ⎜─────⎟⋅⎜───────────⎟              
             ⎝  1  ⎠ ⎜    2      ⎟              
                     ⎝Iₛ⋅s  + c⋅s⎠              
────────────────────────────────────────────────
                          ⎛ ⎛  2 ⎛l₂⋅s    ⎞ ⎞  ⎞
                          ⎜ ⎜-v ⋅⎜──── + 1⎟ ⎟  ⎟
                          ⎜ ⎜    ⎝ v      ⎠ ⎟  ⎟
                          ⎜ ⎜───────────────⎟  ⎟
1   ⎛K_D⋅s⎞ ⎛     1     ⎞ ⎜ ⎝  g⋅(l₁ + l₂)  ⎠  ⎟
─ + ⎜─────⎟⋅⎜───────────⎟⋅⎜────────────────────⎟
1   ⎝  1  ⎠ ⎜    2      ⎟ ⎜      2 ⎛       2  ⎞⎟
            ⎝Iₛ⋅s  + c⋅s⎠ ⎜     s ⋅⎝I_b + h ⋅m⎠⎟
                          ⎜-1 + ───────────────⎟
                          ⎝          g⋅h⋅m     ⎠
```
There is even `1/1` in there. I would have expected that the natural way to represent all of these things is as ratios of polynomials with gcd cancelled. Apparently there was some need to keep common zeros in the numerator and denominator (i.e. not cancelling zeros with poles) but I don't know enough about control theory to understand why that is. Otherwise I would have said that it would probably be better to use `Poly` here rather than `Expr` and then just use `cancel` (but use cancel always rather than having an explicit "simplify" function).

If you do want to cancel gcd from the numerator and denominator then `cancel` is the appropriate function. A lighter weight option is `factor_terms` if you are only pulling out a single factor of `s` from top and bottom.

**Comment by moorepants** [comment]
Yes, we decided to keep the numerator and denominator as distinct expressions, because you may not want automatic simplification and you may want to access the original forms of transfer function. This has some downsides and maybe not as many upsides as we imagined then.

**Comment by moorepants** [comment]
> If you do want to cancel gcd from the numerator and denominator then cancel is the appropriate function. A lighter weight option is factor_terms if you are only pulling out a single factor of s from top and bottom.

Maybe a simple cancel() method would be good to implement then.

**Comment by oscarbenjamin** [comment]
> we decided to keep the numerator and denominator as distinct expressions

The fact that they are distinct expressions is not what I am referring to but rather the fact that factors are not supposed to be cancelled. Supposedly it was important to keep `s/s` rather than have that cancel to 1 because there is a need to prevent pole-zero cancellation. I don't know how important that really is but it prevents representing the transfer function in the natural standard form for rational functions in symbolic computing which is a ratio of polynomials with integer coefficients with polynomial gcd cancelled (i.e. the output of the `cancel` function).

**Comment by sylee957** [comment]
> I don't know how important that really is but it prevents representing the transfer function in the natural standard form for rational functions in symbolic computing which is a ratio of polynomials with integer coefficients with polynomial gcd cancelled

I think that the other asepect I may know, besides the pole-zero cancellation, is the block diagram.
I see that the correspondence of rational forms and block diagram becomes much distant if we automatically simplify them.

However, if I extend this argument,
any connection of transfer functions can just be reduced to a huge single transfer function like
$$\frac{p_{10} s^{10} + ... + p_0}{q_{100} s^{100} + ... + q_0}$$

However, if this is realized in the block diagram, it would just draw like one feedback connection of two very huge parellel connections, which may not be useful.

And also, I would note that it is not easy to convert all rational functions, into block diagram, than vice versa
For example, feedback connection is one to one correspondent to $\frac{F(s)}{1 + Q(s)}$, so once you `expand`, `together` and `cancel` the fractions, it is not easy to get back the original block diagram.

I don't think that it is difficult to do `expand, together, cancel` just before we need heavy calculation. 
I think that the only problem is that the printing rational functions is much more unreadable than helpful,
However, it should be improved to give more readable formats (like pretty printed diagrams, or graphs)

**Comment by moorepants** [comment]
Yes, we wanted to maintain the structure of the block diagram with the transfer function object so that we still have the ability to manipulate it, form variations of transfer functions in it, etc.

**Comment by leomanga** [comment]
From the discussion I gathered that we don't want automatic simplifications in our TransferFunction. However, it's not clear to me whether a function that performs zero-pole cancellations is still acceptable.

I think a function like:
```python
def to_standard_form(self):
    tf = self.doit().expand()
    standard_form = cancel(tf.num / tf.den)
    return TransferFunction.from_rational_expression(standard_form, self.var)
```
might be handy.
